### PR TITLE
[Issue#515] Frontend refactor - Multilayer management part1 

### PR DESCRIPTION
**Pull request by AC-Unicorn**
### Title
Improve frontend with a more general design
part 1 change heatmap control to new desgin

### overview

- common/service.js has been modified to adapt the new design

- sidebar/controllers.js now fires new events to trigger the new design

- map/multi_layer.js is now watching variable and listening events for heatmap

- common/services-multilayer.js as an entry of multilayer factory.

- common/services-multilayer-heatmap.js is providing a general template for any kind(different data resource) of heatmap

- twittermap/main.scala.html now includes all new files

### CODE REVIEW

**Comment by chenlica** [comment]
@AC-Unicorn : a few high-level comments about this PR.  (1) Don't use your name in the subject; (2) If there is an issue, include its number in the subject.  You can check other PRs to get some idea.  (3) Give an overview about this PR.

File: examples/twittermap/web/app/views/twittermap/index.scala.html [hunk 1]
```diff
@@ -42,7 +42,7 @@
         <mapchoose>
           <li role="presentation"><img id="img1" name="img1" title="Count Map" src="/assets/images/aggregation_map.png" width="63" height="63" ></li><br/>
           <li role="presentation"><img id="img2" name="img2" title="Heat Map" src="/assets/images/heat_map_no_border.png" width="63" height="63" ></li><br/>
-          <li role="presentation"><img id="img3" name="img3" title="Pin Map" src="/assets/images/point_map_no_border.png" width="63" height="63"></li>
+          <li role="presentation"><img id="img3" name="img3" title="Pin Map" src="/assets/images/point_map_no_border.png" width="63" height="63"></li>
```
Comment (baiqiushi) [inline examples/twittermap/web/app/views/twittermap/index.scala.html]: Please get rid of format change.

File: examples/twittermap/web/conf/application.conf [hunk 2]
```diff
@@ -105,11 +105,11 @@ endDate = "2017-05-01T00:00:00.000"
 # cacheThreshold = 30000
 
 # for the debug purpose, we can load a smaller json to speed up the front-end
-us.city.path = "/public/data/city.sample.json"
+# us.city.path = "/public/data/city.sample.json"
```
Comment (baiqiushi) [inline examples/twittermap/web/conf/application.conf]: Please do not submit changes to application.conf, since it has nothing to do with your PR.

File: examples/twittermap/web/public/javascripts/common/services-multilayer.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+angular.module('cloudberry.common')
+    .service('createLayerService', function(multilayerHeatmap){
+        var createLayerService = {
+            polygon: {},
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services-multilayer.js]: Save these empty key-values for later PRs.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 4]
```diff
@@ -253,7 +253,10 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
     }
 
     var cloudberryService = {
-
+      
+        
+      layer:{
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: Add a comment for this new variable.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 5]
```diff
@@ -273,6 +276,39 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
       commonTimeSeriesResult: [],
       commonHashTagResult: [],
       errorMessage: null,
+      
+      getFilter: function getFilter(parameters, maxDay, geoIds) {
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: What is this function used for? If no, let's remove it.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 6]
```diff
@@ -368,7 +404,8 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
               ws.send(batchWithPartialGeoRequest);
             }
             break;
-            
+
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: Remove this format change.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 7]
```diff
@@ -488,81 +525,105 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
             ws.send(pointsTimeJson);
             break;
           
+
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: Same.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 8]
```diff
@@ -488,81 +525,105 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
             ws.send(pointsTimeJson);
             break;
           
+          
           default:
             // unrecognized map type
             break;
         }
+
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: Remove useless code instead of comments.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 9]
```diff
@@ -488,81 +525,105 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
...
-          case "heatTime":
-            if(angular.isArray(result.value)) {
-              cloudberryService.commonTimeSeriesResult = result.value[0];
-            }
-            break;
-          case "points":
-            if(angular.isArray(result.value)) {
-              cloudberryService.commonTweetResult = result.value[0].slice(0, defaultSamplingSize - 1);
-              cloudberryService.pinmapMapResult = result.value[0];
+        if (result.category in cloudberryService.parameters.layers){
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/common/services.js]: Add a comment here.

File: examples/twittermap/web/public/javascripts/map/countmap.js [hunk 10]
```diff
@@ -1,5 +1,6 @@
 angular.module('cloudberry.map')
-  .controller('countMapCtrl', function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+  .controller('countMapCtrl', function($scope, $rootScope, $window, $http, $compile, $timeout, cloudberry, leafletData, cloudberryConfig, Cache) {
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/countmap.js]: Remove unnecessary changes for this PR.

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 11]
```diff
@@ -1,5 +1,5 @@
 angular.module("cloudberry.map")
-  .controller("heatMapCtrl", function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+  .controller("heatMapCtrl", function($scope, $rootScope, $window, $http, $compile, $timeout, cloudberry, leafletData, cloudberryConfig, Cache) {
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Same, remove unnecessary changes.

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 12]
```diff
@@ -85,6 +85,7 @@ angular.module("cloudberry.map")
       
       if (!$scope.heatMapLayer){
         $scope.heatMapLayer = $scope.heat;
+
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Same, format change.

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 13]
```diff
@@ -108,6 +109,41 @@ angular.module("cloudberry.map")
...
+    var l = {
+            active:0,
+            init:initheatMap,
+            data:"",
+            draw:drawHeatMap,
+            clear:cleanHeatMap,
+            watchVariables:watchVariables
+    }
+    
+    /*
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Same, remove useless code instead of comments.

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 14]
```diff
@@ -120,13 +156,16 @@ angular.module("cloudberry.map")
       }
     })
     
-    $scope.$watch(
+    /*$scope.$watch(
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Same.

File: examples/twittermap/web/public/javascripts/map/multi_layer.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
+angular.module('cloudberry.map')
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/multi_layer.js]: Better change the file name to `multilayer-manager.js`, which will be more meaningful.

File: examples/twittermap/web/public/javascripts/map/multi_layer.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  .controller('multiLayerCtrl', function($timeout, $scope, $rootScope, $window, $http, $compile, cloudberryConfig, cloudberry, leafletData, Cache, createLayerService) {
+    
+    cloudberry.parameters.layers = {};
+    $scope.watchVariables = {};
+    
+    // initialize
+
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/multi_layer.js]: Please use consistent format for this new file and add some concise comments for each block or function.

File: examples/twittermap/web/public/javascripts/map/pinmap.js [hunk 17]
```diff
@@ -1,5 +1,5 @@
 angular.module("cloudberry.map")
-  .controller("pinMapCtrl", function($scope, $rootScope, $window, $http, $compile, cloudberry, leafletData, cloudberryConfig, Cache) {
+  .controller("pinMapCtrl", function($scope, $rootScope, $window, $http, $compile, $timeout, cloudberry, leafletData, cloudberryConfig, Cache) {
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/map/pinmap.js]: Same, remove unnecessary changes for this PR.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 18]
```diff
@@ -96,6 +97,7 @@ angular.module('cloudberry.sidebar', ['cloudberry.common'])
                 icon2.src = "/assets/images/heat_map_no_border.png";
                 icon3.src = "/assets/images/point_map_no_border.png";
                 $rootScope.$emit("maptypeChange", [premaptype, cloudberry.parameters.maptype]);
+                $rootScope.$emit("multiLayer", [premaptype, cloudberry.parameters.maptype]);
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]: Seems no need to add this new event since it's identical to previous event.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 19]
```diff
@@ -109,6 +111,7 @@ angular.module('cloudberry.sidebar', ['cloudberry.common'])
                 icon2.src = "/assets/images/heat_map.png";
                 icon3.src = "/assets/images/point_map_no_border.png";
                 $rootScope.$emit("maptypeChange", [premaptype, cloudberry.parameters.maptype]);
+                $rootScope.$emit("multiLayer", [premaptype, cloudberry.parameters.maptype]);
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]: Same as above.

File: examples/twittermap/web/public/javascripts/sidebar/controllers.js [hunk 20]
```diff
@@ -122,6 +125,7 @@ angular.module('cloudberry.sidebar', ['cloudberry.common'])
                 icon2.src = "/assets/images/heat_map_no_border.png";
                 icon3.src = "/assets/images/point_map.png";
                 $rootScope.$emit("maptypeChange", [premaptype, cloudberry.parameters.maptype]);
+                $rootScope.$emit("multiLayer", [premaptype, cloudberry.parameters.maptype]);
```
Comment (baiqiushi) [inline examples/twittermap/web/public/javascripts/sidebar/controllers.js]: Same as above.

**Comment by baiqiushi** [comment]
(1) Please take care of the "Codacy/PR Quality Review" suggestions as much as possible.
(2) There are still some previous suggestions not handled.
(3) High level suggestion: when you write comments for code, please add a line on top of that line, instead of add // to the end of current line.
(4) Let's work together on Monday to solve rest problems and also resolve the conflicts.
